src/lib.rs

```diff
@@ -38,6 +39,45 @@ pub trait Database: plumbing::DatabaseStorageTypes {
     {
         <Self as plumbing::GetQueryTable<Q>>::get_query_table(self)
     }
+
+    /// This function is invoked at key points in the salsa
+    /// runtime. It permits the database to be customized and to
+    /// inject logging or other custom behavior.
+    fn salsa_event(&self, event_fn: impl Fn() -> Event<Self>) {
+        #![allow(unused_variables)]
+    }
+}
+
+pub struct Event<DB: Database> {
```

Comment:
#[derive(Debug)] will be useful here

Reply:
Heh, yes, very much so. Also the "other things" (e.g., `Eq`, `Hash`).

Reply:
Ah, I can't really *derive* because of the `DB` parameter. Have to manually implement.

Reply:
Done.